modules/puppetca_token_whitelisting/puppetca_token_whitelisting_autosigner.rb

```diff
@@ -0,0 +1,96 @@
+require 'jwt'
+require 'openssl'
+
+module ::Proxy::PuppetCa::TokenWhitelisting
+  class Autosigner
+    include ::Proxy::Log
+    include ::Proxy::Util
+
+    def sign_all
+      Proxy::PuppetCa::TokenWhitelisting::Plugin.settings.sign_all
+    end
+
+    def hosts_file
+      Proxy::PuppetCa::TokenWhitelisting::Plugin.settings.hosts_file
+    end
+
+    def smartproxy_cert
+      OpenSSL::PKey::RSA.new File.read Proxy::SETTINGS.ssl_private_key.to_s
+    end
+
+    def ensure_hostsfile
+      unless File.exist?(hosts_file)
+        FileUtils.mkdir_p hosts_file.rpartition('/').first
+        FileUtils.touch hosts_file
+        File.write(hosts_file, [].to_yaml)
+      end
+    end
+
+    # Invalidate a token based on the certname
+    def disable certname
+      hosts = autosign_list
+      hosts.delete(certname)
+      File.write hosts_file, hosts.to_yaml
+    end
+
+    # Create a new token for a certname
+    def autosign certname
+      ensure_hostsfile
+      payload = { certname: certname, exp: Time.now.to_i + 10 * 60 * 60 }
+      token = JWT.encode payload, smartproxy_cert, 'RS512'
+      File.write hosts_file, autosign_list.push(certname).to_yaml
+      { generated_token: token }.to_json
+    end
+
+    # List the hosts that are currently valid
+    def autosign_list
+      ensure_hostsfile
+      YAML.load_file(hosts_file).to_a
```

Comment:
Should `safe_load` be used? Especially if the default location is in `/tmp` there's a risk of executing arbitrary code as the `foreman-proxy` user as long as you can write anything in `/tmp`. Because of it I'd actually consider JSON rather than YAML. Since no user has to actually modify the file, it doesn't matter that it's not very human friendly.